elytron/WFCORE-3947-ocsp-ssl-revocation.adoc

```diff
@@ -0,0 +1,81 @@
+= OCSP certificate revocation
+:author:            Martin Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3947[WFCORE-3947]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/ELY-1617[ELY-1617]
+
+=== Dev Contacts
+
+* mailto:mmazanek@redhat.com[Martin Mazanek]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* Security
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* New class RevocatingTrustManager
+* Configurable certificate revocation using OCSP. This will be implemented using PKIXRevocationChecker.
+* Configuration will be done using Builder class and will include:
+** Current configuration in X509CRLExtendedTrustManager
+** OCSP Responder URI for overriding default, where the default is retrieved from AIA extension (RFC5280) or from 'ocsp.responderURL' security property
+** Default revocation mechanism (OCSP/CRL) in a form of prefer crl (default false)
+** Fallback behaviour - cases for unavailable CRL / OCSP responder services
+** Checking of leaf certificates only
+* Deprecating X509CRLExtendedTrustManager, both mechanisms will share one class for fallback purposes and easier configuration
+
+Proposed configuration:
+
+```xml
+<trust-manager name="trust-with-ocsp-crl" algorithm="PKIX" key-store="ElytronCaTruststore">
+    <certificate-revocation-list path="crl.pem" maximum-cert-path="2"/>
+    <ocsp responder="responderUrl"/>
+    <soft-fail/>
+    <only-end-entity/>
+    <prefer-crls/>
+</trust-manager>
```

Comment:
What is the behavior here? I am a little bit confused ATM. Both OCSP and CRL are set - thanks to `perfer-crl`, the CRL check is performed first, in case of failure then the OCSP and in case of another failure thanks to the 'soft-fail' element the request passes, correct?

Is it a `prefer-crl` or a `prefer-crls`?

Reply:
Yes, that is the scenario. Basically if you can't get a response with status of the certificate, the request passes. This is usually the default behaviour of browsers, if they have revocation checking enabled. It is not really secure, but that is why default behaviour would be hard fail, where you will deny the request if you cannot get the status.

prefer-crls, so it reflects the Java API